src/dune_rules/lock_rules.ml

```diff
@@ -2,27 +2,212 @@ open Import
+                    let name = Pkg_workspace.Repository.name repo in
+                    Loc.none, name)
+                in
+                (let open Option.O in
+                 let+ lock_dir = Workspace.find_lock_dir workspace lock_dir_path in
+                 lock_dir.repositories)
+                |> Option.value ~default
+              in
+              let available_repos =
+                List.map workspace.repos ~f:(fun repo ->
+                  Pkg_workspace.Repository.name repo, repo)
+                |> Pkg_workspace.Repository.Name.Map.of_list_exn
+              in
+              let module Repository = Pkg_workspace.Repository in
+              repositories
+              |> Fiber.parallel_map ~f:(fun (loc, name) ->
+                match Repository.Name.Map.find available_repos name with
+                | None ->
+                  User_error.raise
+                    ~loc
+                    [ Pp.textf "Repository '%s' is not a known repository"
+                      @@ Repository.Name.to_string name
+                    ]
+                | Some repo ->
+                  let loc, opam_url = Repository.opam_url repo in
+                  (match OpamUrl.classify opam_url loc with
+                   | `Git -> Opam_repo.of_git_repo loc opam_url
+                   | `Path path ->
+                     Fiber.return @@ Opam_repo.of_opam_repo_dir_path loc path
+                   | `Archive ->
+                     User_error.raise
+                       ~loc
+                       [ Pp.textf
+                           "Repositories stored in archives (%s) are currently \
+                            unsupported"
+                         @@ OpamUrl.to_string opam_url
+                       ]))
+              |> Memo.of_non_reproducible_fiber))
+       and+ pins =
+         Action_builder.of_memo
+           (Memo.of_thunk (fun () ->
+              let open Memo.O in
+              let* project_pins_db = project_pins in
+              let workspace_pins_db =
+                let workspace_pins = Pin.DB.Workspace.of_stanza workspace.pins in
+                let pin_map = Dune_lang.Pin_stanza.Workspace.map workspace.pins in
+                let all_pin_names =
+                  pin_map
+                  |> String.Map.to_list
+                  |> List.fold_left ~init:[] ~f:(fun acc (_repo_name, pkg_map) ->
+                    pkg_map
+                    |> Dune_lang.Package_name.Map.to_list
+                    |> List.fold_left
+                         ~init:acc
+                         ~f:(fun acc (pkg_name, ((loc, _url), _pkg)) ->
+                           (loc, Dune_lang.Package_name.to_string pkg_name) :: acc))
+                in
+                Pin.DB.Workspace.extract workspace_pins ~names:all_pin_names
+              in
+              let combined_pins = Pin.DB.combine_exn workspace_pins_db project_pins_db in
+              Memo.return combined_pins
+              >>| resolve_project_pins
```

Comment:
How is this code not shared with the non watch mode locking code? Why would we handle pins differently here?